espresso/environment/2_espresso_liveness_test.go

```diff
@@ -119,3 +131,285 @@ func TestE2eDevNetWithEspressoEspressoDegradedLiveness(t *testing.T) {
+			if have, want := err, error(nil); have != want {
+				t.Fatalf("Waiting for L2 tx:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+			}
+
+			// We not have a receipt from the L2 Sequencer, indicating that
+			// the transaction was successfully included in a block.
+			received := time.Now()
+
+			submissions = append(submissions, submission{
+				receipt:   receipt,
+				created:   created,
+				submitted: submitted,
+				received:  received,
+			})
+		}
+
+		// Let's verify that all of our transactions came through successfully,
+		// using our Caff Node as the verification client.
+		for i, submission := range submissions {
+			receipt, err := wait.ForReceiptOK(ctx, caffVerif, submission.receipt.TxHash)
+			if have, want := err, error(nil); have != want {
+				t.Fatalf("Waiting for L2 tx on verification client:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+			}
+
+			// Transaction Hash should match
+			if have, want := receipt.TxHash, submission.receipt.TxHash; have != want {
+				t.Errorf("Receipt tx hash mismatch for submission %d:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", i, have, want)
+			}
+
+			// Block Hash should match
+			if have, want := receipt.BlockHash, submission.receipt.BlockHash; have != want {
+				t.Errorf("Receipt block hash mismatch for submission %d:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", i, have, want)
+			}
+		}
+
+		// Alice's balance should have increased by N
+		balanceAliceFinal, err := caffVerif.BalanceAt(ctx, addressAlice, nil)
+		if have, want := err, error(nil); have != want {
+			t.Fatalf("Failed to fetch Alice's balance:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+		}
+
+		expectedBalance := new(big.Int).Add(balanceAliceInitial, big.NewInt(int64(N)))
+		if balanceAliceFinal.Cmp(expectedBalance) != 0 {
+			t.Errorf("Alice's balance did not increase as expected:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", balanceAliceFinal, expectedBalance)
+		}
+	}
+
+	// Tell the Streamer to stop streaming.
+	streamBlocksCancel()
+	wg.Wait()
+
+	if have, want := len(espressoReceipts), N; have < want {
+		t.Fatalf("Expected to received at least many batches as submissions:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+	}
+
+	// We'll check that our timings meet or exceed the requirements of the test.
+	var totalDiff time.Duration
+	var totalDenom time.Duration
+	for i, submission := range submissions {
+		espressoReceived, ok := espressoReceipts[submission.receipt.BlockHash]
+		if have, want := ok, true; have != want {
+			t.Errorf("Failed to find batch for submission %d:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", i, have, want)
+			continue
+		}
+
+		diff := espressoReceived.received.Sub(submission.received)
+		totalDiff += diff
+		totalDenom++
+
+		if have, want := diff, 10*time.Second; have > want {
+			t.Errorf("Submission %d was not confirmed in an espresso block within 10 seconds of submission:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", i, diff, want)
+		}
+	}
+
+	if totalDenom > 0 {
```

Comment:
Should not we require that `totalDenom > 0`?

Reply:
Sure, there's no harm in doing this check. If we got to this point with `totalDenom` being `0`, then we would have already errored.  But again, no harm in adding another check.